StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/IntentRepresentable.swift

```diff
@@ -0,0 +1,109 @@
+//
+//  IntentRepresentable.swift
+//  StripePaymentSheet
+//
+//  Created by Nick Porter on 2/13/23.
+//
+
+import Foundation
+@_spi(STP) import StripePayments
+
+protocol IntentRepresentable: PaymentMethodRequirementProvider {
```

Comment:
In the future our new "fake" intent will conform to this protocol, as `Intent` now does.

Reply:
Dropping some half baked thoughts early - I'm starting to think it's going to be useful to express this as an Enum with two cases like `.intent` and `.intentless` - when it comes time to confirm, we're going to want to explicitly switch over these two cases and do different things.

`Intent` is already an enum of paymentIntent and setupIntent - what if we added a third called `deferredIntent` or something?

Reply:
I really like the idea of an enum as you described, I quickly put something together to see what it would look like and one downside I'm seeing is that if we add a `deferredIntent` case to `Intent` we have to switch over it in many different places that would result in duplication of code and possibly redundant control flow. 

For example in `PaymentSheetViewController` when building the `buyButton` we would have to switch over an additional `deferredCase` which then we would need to add some logic to determine if it is a payment or setup deferred intent and duplicate the logic that is already in the existing two cases.

```swift
switch intent {
case .paymentIntent(let paymentIntent):
    return .pay(amount: paymentIntent.amount, currency: paymentIntent.currency)
case .setupIntent:
    return .setup
case .deferredIntent(let deferredIntent):
   if deferredIntent.mode == .setup {
      return .setupIntent
   } else 
      return .pay(amount: paymentIntent.amount, currency: paymentIntent.currency)
   }
}
```

There are some other cases where larger chunks of code are within each switch case that would be a bit uglier. My goal was to make it so PaymentSheet (mainly the UI) has no concept or understanding of a deferred intent so that our control flow when handling an intent can remain the same throughout the codebase except for when we confirm the intent and maybe a few other spots. Maybe there is some fancier switching we could do with the enums that would improve this? But in general I'm hoping we can have files like PaymentSheetFormFactory not need to know about a deferred intent.

Let's chat in our pairing session tomorrow!

Reply:
Some raw notes ahead of pairing:

I think in this specific case I prefer the explicitness over using amount and currency as a proxy for payment vs setup mode. e.g. it's possible that SI will contain amount and currency in the future (there is already a proposal to add currency to it).  

But I get your point in the general. I think we can kind of combine both worlds - add `deferredIntent` _and_ add vars that are in `IntentRepresentable` like `setupFutureUsage`, `recommendedPaymentSheetMethodTypes` etc. to `Intent` directly to avoid switching unless it's really necessary.  In the example above, perhaps we add a `var mode: Payment or Setup` to `Intent` to help with that.